src/main/java/com/google/sps/ImageTextDectector.java

```diff
@@ -16,45 +16,109 @@
 
 import com.google.sps.data.ShoppingListTextEntry;
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
 
 /**
- * This class generates a shopping list based on image 
- * 1) It uses cloudVisionAPI to scan an image containing shopping
+ * This class generates a shopping list based on image provided in {@link #extractShoppingList(byte[])}
+ * 1) It uses cloudVisionAPI to scan the image containing shopping
  * list items and detect text from it. 
- * 2) It then uses an algorithm to create shopping sentences (queries)
- * from the text and their position. 
- * 3) This list of queries is returned to the Servlet from the method 'extractShoppingList'.
+ * 2) It then uses an algorithm to extract shopping sentences (queries)
+ * from the text based on their position (y coordinates). 
+ * 3) This list of queries is returned to the Servlet from the {@link #extractShoppingList(byte[])}.
  */
 public class ImageTextDectector {
 
-  TextDetectionAPI textDetectionAPI;
+  // This variable is used to invoke interface's implimentation for real CloudVisionAPI calls and fake calls from tests.
+  private TextDetectionAPI textDetectionAPI;
 
   public ImageTextDectector(TextDetectionAPI textDetectionAPI) {
     this.textDetectionAPI = textDetectionAPI;
   }
 
-  public String extractShoppingList(byte[] shoppingImageBytes)
+  public List<String> extractShoppingList(byte[] shoppingImageBytes)
       throws IOException, PhotoDetectionException {
     List<ShoppingListTextEntry> shoppingListText = textDetectionAPI.detect(shoppingImageBytes);
 
-    return createShoppingListQuery(shoppingListText);
+    return createShoppingListQueries(shoppingListText);
   }
 
-  /** Creates query from the text detected by cloudVision API. */
-  private String createShoppingListQuery(List<ShoppingListTextEntry> shoppingListText)
+  /** Creates query from the text detected by Cloud Vision API. */
+  private List<String> createShoppingListQueries(List<ShoppingListTextEntry> shoppingListText)
       throws PhotoDetectionException {
-    // ToDo: Make an algorithm to create query sentences by separating out text returned by
-    // cloudVisionAPI; to group shoppping items based on their position (y axis).
     if (shoppingListText.isEmpty()) {
       throw new PhotoDetectionException("Shopping List doesn't contain any text");
     }
 
-    String queryItem = "";
-    for (ShoppingListTextEntry singleWord : shoppingListText) {
-      queryItem += singleWord.getText() + " ";
+    // Remove the first element of this list because cloud vision API returns all the 
+    // text detetcted as the first element of the list followed by list of single 
+    // words and their properties. Examples of text returned from API can be seen in the test file.
+    shoppingListText.remove(0);
```

Comment:
Is this documented somewhere in the API why the additional string is returned? If so, please link to the documentation.

I'm just wondering if there are cases where, for example to text blocks are found and they are returned as two full strings + the individual words.

Reply:
Below are the pages from Google doc with sample response. It took me a while to search for these; as they are not clearly visible on the website.
1) There is hidden "Response" section on Detect Text in Images page here : https://cloud.google.com/vision/docs/ocr#detect_text_in_a_remote_image
2) I also found a similar response for the TextAnnotationImagesOffline method here : https://cloud.google.com/vision/docs/batch#offline_batch_outputoutput-1-to-2json

I validated it with the output generated in my logs while debugging. Based on my findings (below), I wrote the code to extract the words.
1) Vision API first returns the entire text from the image in a single line as first "description": keyword under under the keyword "textAnnotations": section 
I am ignoring this line in my code as it has coordinates for the entire image and not word boxes.

2) Then in subsequent  "description": : sections, the Vision API starts returning coordinates of each individual word box starting from lower x and lower y values which is bottom left on the axis.

I've added these URLs in the documentation.

Reply:
Thanks 👍